### ISSUE TITLE
What's the recommended way to load a snippet for all file types except a few?

### ISSUE THREAD

**Issue by peey** [SirVer/ultisnips#1059]
Here's what I have tried

```
fun! LoadSelectedSnippets()
  " Don't load on these file types
  if &ft =~ 'php\|ruby'
    unlet g:UltiSnipsSnippetsDir " by default this is unset
    let g:UltiSnipsSnippetDirectories=["UltiSnips"] " this is its default value
    call UltiSnips#RefreshSnippets()
  else 
    let g:UltiSnipsSnippetsDir="~/.vim/mysnippets/"
    let g:UltiSnipsSnippetDirectories=["UltiSnips", "mysnippets"]
    call UltiSnips#RefreshSnippets()
  endif
endfun

autocmd BufEnter * call LoadSelectedSnippets()
```

Where `~/.vim/mysnippets/` contains `all.snippets`. 

I'm aware this could be optimized to remove calls to `RefreshSnippets` by setting `b:` variants instead of `g:` variants. Nonetheless this solution only lets you remove _all_ user defined snippets. It doesn't work if my requirements are one set of snippets in `allButPhpAndRuby.snippets` and another set of snippets in `all.snippets`.

Even scripting using the provided function `UltiSnips#AddSnippetWithPriority` isn't sufficient because there is no way to remove a snippet.

**Comment by peey** [comment]
I found a hack. I'd still wait on maintainers' word on what's the recommended way to do this. Also I'd suggest that `UltiSnipsAddFiletypes` should be renamed to `UltiSnipsSetFiletypes` or perhaps be abolished altogether in favor of a variable which you set using normal set command

```
" Working principle: use a fake filetype. Snippets are in
" faketype.snippets. Append the type to the current using 
" UltiSnipsAddFiletypes when arbitary conditions are satisfied. Here the
" arbitary condition is that it shouldn't be one of the filetypes to exclude
let g:UltiSnipsSnippetsDir="~/.vim/mysnippets"
let g:UltiSnipsSnippetDirectories=["UltiSnips", "mysnippets"]
fun! LoadSelectedSnippets()
  " Don't strip on these filetypes
  if &ft =~ 'tex\|plaintex\|pandoc'
    " do nothing
  else 
    if strlen(&ft)
      execute 'UltiSnipsAddFileypes' &ft . "." . "faketype"
    else
      execute 'UltiSnipsAddFiletypes faketype'
    endif
  endif
endfun

autocmd BufEnter * call LoadSelectedSnippets()

**Comment by SirVer** [comment]
UltiSnips is designed with whitelisting in mind, i.e. you are supposed to add to all the filetypes that you work with which snippets you want loaded there. Am I correct in understanding that this a request for blacklisting, e.g. load `foo` for everything, but not `latex` files? If so, I think the workaround you've found is probably among the easiest solutions.

**Comment by SirVer** [comment]
Unpacking this, there are two requests here: 

- Rename and/or get rid of UltiSnipsAddFiletypes. While I (now) agree that a variable would have been the better UI choice instead of a command for this functionality, I do not consider it so broken that it is worthwhile doing the breaking change of removing it at this point in time.
- Blacklisting snippets, i.e. use "all but these filetypes" in a file. While I think this is an interesting feature, I feel it is fairly fringe and I do not see an implementation that is not hard to misunderstand (i.e. how does whitelisting and blacklisting configuration interplay? How to configure it?). At this point, I do not feel this is plugging a large enough hole in the feature list of UltiSnips. If you feel strongly about this, I suggest opening a new issue including a sketch of how the UI of this feature could be implemented.

with these thoughts, I am closing this as non-actionable at the moment. If you disagree, let's reopen and discuss further.
